File: LeetCode/Python/155.min-stack.py
# @before-stub-for-debug-begin
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=155 lang=python3
#
# [155] Min Stack
#

# @lc code=start
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minElement = None
        self.secondMinElement = None


    def push(self, x: int) -> None:
        if self.minElement is None:
            self.minElement = [x, 0]
        
        elif self.minElement[0] > x:
            # self.secondMinElement = self.minElement
            self.minElement = [x, len(self.stack)]
        
        self.stack.append(x)


    def pop(self) -> None:
        self.stack.pop()
        if len(self.stack) == self.minElement[1]:
            if self.stack:
                self.minElement = self.findMin()
            else:
                self.minElement = None


    def top(self) -> int:
        return self.stack[-1]


    def getMin(self) -> int:
        return self.minElement[0]
    
    def findMin(self) -> List[int]:
        minElement = [self.stack[0], 0]
        for i in range(1, len(self.stack)):
            if self.stack[i] < minElement[0]:
                minElement = [self.stack[i], i]
        
        return minElement
# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
# @lc code=end
def main():
    obj = MinStack()
    obj.push(-2)
    obj.push(0)
    obj.push(-3)
    print(obj.getMin())
    obj.pop()
    print(obj.top())
    print(obj.getMin())

main()

File: LeetCode/CPP/155.min-stack.cpp
/*
 * @lc app=leetcode id=155 lang=cpp
 *
 * [155] Min Stack
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Node
{
public:
    int val;
    int min;
    Node *next;
    Node() : val(0), min(0), next(nullptr) {}
    Node(int val, int min) : val(val), min(min), next(nullptr) {}
    Node(int val, int min, Node *next) : val(val),
                                         min(min),
                                         next(next) {}
};
class MinStack
{
public:
    Node *head;
    MinStack()
    {
    }

    void push(int val)
    {
        cout << "Pushing " << val;
        if (head == nullptr)
            head = new Node(val, val);
        else
        {
            Node *tmp = new Node(val, min(val, head->min), head);
            head = tmp;
        }
        cout << "Pushing " << val;
    }

    void pop()
    {
        cout << "Popping " << head->val;

        head = head->next;
    }

    int top()
    {
        cout << "Topping " << head->val;
        return head->val;
    }

    int getMin()
    {
        cout << "Min Ele " << head->min;
        return head->min;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
// @lc code=end
